### PR TITLE
Fix thread safety of RegularizedForwardPass

### PR DESCRIPTION
**Pull request by odow**
Closes https://github.com/odow/SDDP.jl/issues/791

### CODE REVIEW

File: src/plugins/forward_passes.jl [hunk 1]
```diff
@@ -353,20 +355,45 @@ function forward_pass(
...
+    #  * It might also be that we "skip" some of the starting tirla points,
+    #    because Thread A sets a trial centre, then thread B comes along and
+    #    sets a new one before A can start the forward pass. Again, this doesn't
+    #    really matter; this regularization is just a performance optimization.
+    lock(node.lock) do
+        for (k, (l, u)) in fp.old_bounds
+            fp.trial_centre[k] = pass.sampled_states[1][k]
+            set_lower_bound(node.states[k].out, l)
+            set_upper_bound(node.states[k].out, u)
+        end
```
Comment (odow) [inline src/plugins/forward_passes.jl:396]:
```suggestion
        end
        return
```

**Comment by odow** [comment]
@FSchmidtDIW try with this branch:

```julia
] add SDDP#od/thread-safety
```

**Comment by FSchmidtDIW** [comment]
@odow Thank you for implementing this!

I guess you also need to get rid of this part of algorithm.jl:

https://github.com/odow/SDDP.jl/blob/aa0cf1b329f6de7fdbba23cb5dd66d7d97faeb40/src/algorithm.jl#L1138-L1142

File: src/plugins/forward_passes.jl [hunk 2]
```diff
@@ -353,20 +355,46 @@ function forward_pass(
...
+    #
+    #  * Thread A might have obtained the lock, modified the bounds to the trial
+    #    centre, released the lock, and then entered `forward_pass`
+    #  * But before it can re-obtain a lock on the first node to solve the
+    #    problem, Thread B has come along below and modified the bounds back to
+    #    their original value.
+    #  * This means that Thread A is solving the unregularized problem, but it
+    #    doesn't really matter because it doesn't change the validity; this is a
+    #    performance optimization.
+    #  * It might also be that we "skip" some of the starting tirla points,
```
Comment (odow) [inline src/plugins/forward_passes.jl:387]:
```suggestion
    #  * It might also be that we "skip" some of the starting trial points,
```

**Pull request merged**